### ISSUE TITLE
Allow declaration `declare` within conditionals

### ISSUE THREAD

**Issue by ganeshkbhat** [Microsoft/TypeScript#25570]
## Search Terms
declare and variable declaration (for issues)

## Suggestion
Allow declaration inside conditionals like below:

```
if (conditional) {
     declare var variable: any | <T>
} 
```

## Use Cases
Lets take an example of Web Worker support from browser. If the Web Worker (window.Worker) is not available I will switch to a polyfill that creates a window.PseudoWorker object option. The pseudoworker is always loaded irrespective of browser support.

In my .ts class, I will create the following:

```
declare var Worker: Worker | undefined;
if (!Worker) {
     declare var PseudoWorker: any; // this fails
}

class y {
  worker: any;
  constructor() {
    if (!Worker) {
         this.worker = new PseudoWorker('myurl.js');
    } else {
         this.worker = new Worker('myurl.js');
    }
  }
}
```

An alternative way to approach the following is:

`declare var PseudoWorker: Worker | any;`

But the issue with this is, I am loading a polyfill with PseudoWorker all the time in the global context irrespective of the fact that Web Worker is supported or not (window.Worker object present or not). Since the Object is always going to be available there is a clash in the object names used.

This is however different from the [https://github.com/Microsoft/TypeScript/issues/23602](https://github.com/Microsoft/TypeScript/issues/23602) issue by the way that:
* The declaration in the global context stays for pseudoworker and worker for supported browsers OR just pseudoworker for unsupported browsers
* The declaration of pseudoworker object to be used inside .ts file module should be optional based on presence of an object based on browser support
* There will be global naming issues (unexpected overriding) if the conditional declaration is not there (incase of pseudoworker when the browser worker support or window.Worker object is available)

An alternate recommendation based on the 23602 issue may be 

`declare? (var Worker: Worker) : (var PseudoWorker: any);` (updated from `declare? var Worker: Worker || var PseudoWorker: any;`)

My suggestion meets these guidelines:
* [ ] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by mhegazy** [comment]
Looks like a duplicate of https://github.com/Microsoft/TypeScript/issues/819 and similar use case as #23602.

Seems like the request is to just allow an override for a variable or a property declaration. correct?

**Comment by ganeshkbhat** [comment]
No its not a duplicate of #819 . Its a different use case. In case of #23602 could be same. However, #23602 addresses only creation/declaration of the variable in case the object is not present. 

Here, in this use case (conditional declaration), I am requesting declaration of a variable if case 1 and declaration of another variable in case 2. 

This is for `declare? (var Worker: Worker) : (var PseudoWorker: any);`

The second use case of allowing conditional declaration is to allow declaration based on a dependent case - declaration of a variable if case 1 and declaration of another variable in case 2 where case 1 object is not defined.

```
declare var Worker: Worker | undefined;
if (!Worker) {
  declare var PseudoWorker: any;
}
```

**Comment by mhegazy** [comment]
i am not sure i understand what you mean by "variable if case 1 and declaration of another variable in case 2." are you looking for dependent types?

**Comment by ganeshkbhat** [comment]
`declare var variable` within a conditional (case) block (case)

**Comment by ganeshkbhat** [comment]
`declare var variable` within a conditional block (case). You mày have dependant types or dependant conditions. Both ways. That's what makes me believe `declare? (var x:<T>) : (var y:<T>)` might solve both cases of selective declaration (eg: #23602 based on type) as well as conditional/block scoped declaration (eg: #25570 based on object not 'necessarily' type). Now, I think we can add this issue's use case #25570 to #23603 to make a new proposal to meet the mentioned use cases.

My thoughts (thinking aloud):

`"declaration of a variable if conditional case 1 (type or global scope object) and declaration of another variable in conditional case 2 (type or global scope object)"`

* Use case:
if (xObj === 1) {
     declare var a: <T>;
} else if (yObj === 2) {
     declare var b: <T>;
}
* Proposed:
Allow conditional declaration as above.
```
if (xObj === 1) {
     declare var a: <T>;
} else if (yObj === 2) {
     declare var b: <T>;
}
```


* Use case:
```
if (xObj === 1) {
     declare var a: <T>;
} else {
     declare var b: <T>;
}
```
* Proposed:
Allow conditional declaration as above.
```
if (xObj === 1) {
     declare var a: <T>;
} else {
     declare var b: <T>;
}
```


* Use case:
```
if (<SOMETYPE>) {
     declare var a: <SOMETYPE>;
} else if (!<SOMETHIRDTYPE>) {
     declare var b: <SOMEFOURTHTYPE> | <T>;
}
```
* Proposed:
```
declare? (var a: <SOMETYPE>) : false;
if (!<SOMETHIRDTYPE>) {
    declare var b: <SOMETHIRDTYPE> | <T>;
}
```
OR
```
if (<SOMETYPE>) {
     declare var a: <SOMETYPE>;
} else if (!<SOMETHIRDTYPE>) {
     declare var b: <SOMEFOURTHTYPE> | <T>;
}
```


* Use case:
```
if (<SOMETYPE>) {
     declare var a: <SOMETYPE>;
} else if (<SOMETHIRDTYPE>) {
     declare var b: <SOMETHIRDTYPE>;
}
```
* Proposed:
```
declare? (var a: <SOMETYPE>) : false;
declare? (var b: <SOMETHIRDTYPE>) : false;
```
OR
```
if (<SOMETYPE>) {
     declare var a: <SOMETYPE>;
} else if (<SOMETHIRDTYPE>) {
     declare var b: <SOMETHIRDTYPE>;
}
```


`declaration of a variable if conditional case 1 (type or global scope object), and declaration of another variable in conditional case 2 (type or global scope object) where conditional case 1 object/type is not defined.`


* Use case:
```
if (<SOMETYPE>) {
     declare var a: <SOMETYPE>;
} else {
     declare var b: <SOMEOTHERTYPE>;
}
```
* Proposed:
`
declare? (var a: <SOMETYPE>) : (var b: <SOMEOTHERTYPE>)
`



* Use case:
```
declare var a: <SOMETYPE> | undefined;
if (!a) {
     declare var b: <SOMEOTHERTYPE> | <T>;
} 
```
* Proposed:
```
declare var a: <SOMETYPE> | undefined;
if (!a) {
     declare var b: <SOMEOTHERTYPE> | <T>;
}

```


* Use case:
```
// Some global OR local object a
if (!aGobalORlocalObject) {
     declare var b: <SOMEOTHERTYPE> | <T>;
} 
```
* Proposed:
Allow conditional declaration as above.
```
// Some global OR local object a
if (!aGobalORlocalObject) {
     declare var b: <SOMEOTHERTYPE> | <T>;
} 
```


`This is different from #23602 where `"selective declaration is made based on presence of type or object"``


Use Case in #23602:
```
if (!<SOMETYPE>) {
     declare var a: <SOMETYPE>;
}
```
AND
```
declare var a: <SOMETYPE> | undefined; // a is declared
```
Proposed in #23602:

```
declare? var a: <SOMETYPE>; // a should not be declared
```

**Comment by kitsonk** [comment]
It sounds like you are trying to solve #21316 a different way.  How would #21316 not solve the problems you are talking about?

**Comment by ganeshkbhat** [comment]
Yes, quite close. But I believe i#216316 does not address some above use cases. Say, I dont have an interface or a type and it is an object presence or availability I am checking for in the global's object. 


Here, lets say 
* we are checking if the Worker Type/Object exists then we assign Worker to declared variable `Worker` else we assign the type any. This case the declared variable `Worker` may have `Worker` object assigned during runtime and `PseudoWorker` object will never be declared and assigned to any variable. In case it is needed, we will have compiletime errors.
* we are checking if the Worker Type/Object exists then we assign Worker to declared variable `PseudoWorker` else we assign any (#21316).  This case declared variable `PseudoWorker` may have `Worker` object assigned during runtime or `PseudoWorker` object to be assigned during runtime based on whether the PseudoWorker polyfill gets loaded before or very late in the app initialization.
* we are checking if the Worker Type/Object exists then we assign Worker to declared variable `Worker` else we assign undefined to Worker and declare `PseudoWorker` with type any. This case allows the globally named PseudoWorker object, which is a polyfill object, to take the value for PseudoWorker.


Would the issue #21316 suffice? I believe not. The first two will be dependent on whether the polyfill object was loaded or not - when the browser supports Worker Object (and polyfill is not needed). Second, in #21316 a conditional type assignation also does not allow for not declaring a variable at all if it is supposed to be undefined (#23602 declare? var x: Worker | undefined). It also does not address the proposal of selectively declaring a variable `Worker` or `PseudoWorker` based on any other condition like thirdVariableObject definition or meeting a condition, OR on a condition where Worker object is available in the browser.

Hence the need for proposals:

`
declare? var Worker: Worker;
`
AND
`
declare? var PseudoWorker: Worker? undefined : any;
`
AND
```
declare var Worker: Worker | undefined;
if (!Worker) {
     declare var PseudoWorker: any;
}
```
AND 
declare? (var Worker: Worker) : (var PseudoWorker: any);

**Comment by mhegazy** [comment]
Looks like all these use cases boil down to conditional types. let's not get stuck on the syntax or how to declare these variables. This request needs the type system to process and understand dependent types (possibly infer them from declarations/conditional statements).

**Comment by mhegazy** [comment]
I do not think we have a specif issue tracking dependent types at the moment. there has been discussions about them in different issues. if you want to log a new suggestions for [dependent types](https://en.wikipedia.org/wiki/Dependent_type) that would be great.

**Comment by ganeshkbhat** [comment]
Yes, I understand. That is something I have not considered in a holistic manner, frankly, while creating the issue. My bad. I was only considering simple conditionals for declaration (whereas dependency types or related values may take different structure altogether) since it was needed for a context of a project. May be you could put your thoughts as well based on requests and your experience. Technically #21316 is a dependency type case; and #25570 's main proposal of conditional block declaration may not really face the same complexity, and allows for more flexibility during creation of the language definition without a lot of use case checks (again thats my view).

The only dependency I was considering was presence or global or local definitions or conditions. That could be fairly easy to take forward then a bunch of dependency value or type checks. The main reason why my main request was actually allowing conditional block declaration. Something like this:

Use case #23602 
```
if (!!Window) {
     declare var Window: Window;
} 
```

Use case #25570

```
if (!!Worker) {
     declare var Worker: Worker;
} 
```
OR
```
declare var Worker: Worker | undefined;
if (!Worker) {
     declare var PseudoWorker: any;
} 
```

OR

```
if (!window.Worker) {
     declare var PseudoWorker: any;
} 
```
OR
```
if (a === 12) {
     declare var PseudoWorker: any;
} 
```
OR
```
// Use case dependency type #21316 declare var Worker: Worker ? Worker : any;
// Can be rewritten as below
declare var Worker: Worker: undefined;
if (!Worker) {
     declare var PseudoWorker: any;
} 
// However this does not resolve the problem of not declaring a variable Worker after all
// Which is resolved by #23602 by not declaring them at all - declare? var Worker:Worker;
// Which is extended by dependency type inception declare? (var PseudoWorker: Worker) : (var PseudoWorker: any). The value is prone for unneeded overriding and incorrect behavior which is the complexity we discussed 
// and better declared as declare? (var Worker: Worker) : (var PseudoWorker: any). 
// Or as mentioned in #21316 like this
// declare var PseudoWorker: any; declare var Worker:Worker ? Worker: PseudoWorker;
// The ways of implementing my referred use case can be done in different ways but other use cases may be left out for dependency types/values as mentioned in other issues.
// However, this does not change the language implementation for dependency types. So, all these use cases will have to be considered holistically to implement multiple declaration ways rather than one single implementation for different/all issue numbers.
// I can raise a separate issue for this if you believe you can consider this as a feature for future. But creating a single issue for dependency type may be oppressive than beneficial (considering we will implement it). Dependency types should actually be a tag, is what I believe. 
// I can for sure create an issue for conditional declation alternative for declare? (var Worker:Worker) : (var PseudoWorker: any)
```

I definitely believe dependency of types can be difficult to implement in different test cases. But conditional block declaration may not be as complex. What are your thoughts on the same?

**Comment by ganeshkbhat** [comment]
Here is another reason why conditional types #21316 will be different from #25570:

```
// Usage in classes and functions where as #25570 just addresses the declaration part.

type DeepReadonly<T> =
    T extends any[] ? DeepReadonlyArray<T[number]> :
    T extends object ? DeepReadonlyObject<T> :
    T;

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}

type DeepReadonlyObject<T> = {
    readonly [P in NonFunctionPropertyNames<T>]: DeepReadonly<T[P]>;
};

function f10(part: DeepReadonly<Part>) {
    let name: string = part.name;
    let id: number = part.subparts[0].id;
    part.id = part.id;  // Error
    part.subparts[0] = part.subparts[0];  // Error
    part.subparts[0].id = part.subparts[0].id;  // Error
    part.updatePart("hello");  // Error
}
```
